# Obtaining /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MPRAOneDesign/mouseCtxLiverSingleRepPeaksMerge50NoProtSlop200.bed:
zcat /projects/pfenninggroup/mouseCxStr/TissueATACseq/Kundaje_Outputs/1_082317C_CTX/peak/macs2/rep1/082317C_CTX_S1_R1_001.trim.PE2SE.nodup.tn5.pf.filt.narrowPeak.gz /projects/pfenninggroup/mouseCxStr/TissueATACseq/Kundaje_Outputs/4_5_082417C_CTX/peak/macs2/pooled_rep/082417C_CTX1_S4_R1_001.trim.PE2SE.nodup.tn5_pooled.pf.filt.narrowPeak.gz /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MouseDNase/Cortex_AgeB_ATAC_out/peak/macs2/rep1/Cortex_ATAC_AgeB_Rep1_techRepsPooled.nodup.tn5.pf.pval0.01.300K.filt.narrowPeak.gz /projects/pfenninggroup/mouseCxStr/TissueATACseq/Kundaje_Outputs/10_16_082917BCTX2_090417BCTX/peak/macs2/rep2/090417B_CTX2_S6_R1_001.trim.PE2SE.nodup.tn5.pf.filt.narrowPeak.gz /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MouseDNase/Liver_ATAC_out/peak/macs2/rep1/ENCFF146ZCO.nodup.tn5.pf.filt.narrowPeak.gz /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MouseDNase/Liver_ATAC_out/peak/macs2/rep2/ENCFF109LQF.nodup.tn5.pf.filt.narrowPeak.gz | sort -u -k1,1 -k2,2n -k3,3n | mergeBed -d 50 -i stdin | subtractBed -a stdin -b /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MouseGenome/gencode.vM15.annotation.protTranscript.bed | slopBed -i stdin -g /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MouseGenome/mm10.chrom.sizes -b 200 | sort -u -k1,1 -k2,2n -k3,3n > /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MPRAOneDesign/mouseCtxLiverSingleRepPeaksMerge50NoProtSlop200.bed

# Obtaining /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MPRAOneDesign/mouseCtxLiverSingleRepPeaksMerge50NoProtSlop200_ctxUp.bed:
paste /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MPRAOneDesign/mouseCtxLiverSingleRepPeaksMerge50NoProtSlop200.bed /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MPRAOneDesign/CtxVsLiver.csv | awk 'BEGIN{OFS="\t"} $5 > 1' | awk 'BEGIN{OFS="\t"} $9 < 0.05' > /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MPRAOneDesign/mouseCtxLiverSingleRepPeaksMerge50NoProtSlop200_ctxUp.bed

# Obtaining /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MPRAOneDesign/mouseBrainMinusGpeLiver_crossAge_allBrainMinusGpeUp.bed:
intersectBed -wa -a /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MPRAOneDesign/mouseCtxLiverSingleRepPeaksMerge50NoProtSlop200_ctxUp.bed -b /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MPRAOneDesign/mouseStrLiverSingleRepPeaksMerge50NoProtSlop200_strUp.bed | cut -f1-3 | sort -u -k1,1 -k2,2n -k3,3n > /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MPRAOneDesign/mouseBrainMinusGpeLiver_crossAge_allBrainMinusGpeUp.bed

# Obtaining /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MPRAOneDesign/mouseBrainMinusGpeLiver_crossAge_allBrainMinusGpeUp_enhancer_backgroundDinuc.txt:
closestBed -d -t first -a /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MPRAOneDesign/mouseBrainMinusGpeLiver_crossAge_allBrainMinusGpeUp.bed -b /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MouseGenome/gencode.vM15.annotation.protTranscript.bed | awk 'BEGIN{OFS="\t"} $7 > 20000' | cut -f1-3 | fastaFromBed -fi /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MouseGenome/mm10.fa -bed stdin -fo /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MPRAOneDesign/mouseBrainMinusGpeLiver_crossAge_allBrainMinusGpeUp_enhancer.fa
fasta-get-markov -m 1 -dna /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MPRAOneDesign/mouseBrainMinusGpeLiver_crossAge_allBrainMinusGpeUp_enhancer.fa /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MPRAOneDesign/mouseBrainMinusGpeLiver_crossAge_allBrainMinusGpeUp_enhancer_backgroundDinuc.txt

# Obtaining /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_hg38.bed
fimo --bgfile /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MPRAOneDesign/mouseBrainMinusGpeLiver_crossAge_allBrainMinusGpeUp_enhancer_backgroundDinuc.txt --max-stored-scores 5000000 --motif M4467_1.02 --thresh 0.001 --o /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MPRAOneDesign/mouseBrainMinusGpeLiver_crossAge_allBrainMinusGpeUp_enhancer_Mef2cMotifHits /home/ikaplow/RegulatoryElementEvolutionProject/MotifData/motif_databases/CIS-BP/HumanMacaqueMouseFinchFilt.meme /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MPRAOneDesign/mouseBrainMinusGpeLiver_crossAge_allBrainMinusGpeUp_enhancer.fa
python /home/ikaplow/RegulatoryElementEvolutionProject/src/RegElEvoCode/convertFIMOToMotifHitBed.py --fimoFileName /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MPRAOneDesign/mouseBrainMinusGpeLiver_crossAge_allBrainMinusGpeUp_enhancer_Mef2cMotifHits/fimo.txt --nameCol 1 --pValCutoff 0.0001 --outputFileName /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MPRAOneDesign/mouseBrainMinusGpeLiver_crossAge_allBrainMinusGpeUp_enhancer_Mef2cMotifHits/fimo.bed
intersectBed -wa -wb -a /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt.narrowPeak.gz -b /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MPRAOneDesign/mouseBrainMinusGpeLiver_crossAge_allBrainMinusGpeUp_enhancer_Mef2cMotifHits/fimo.bed | sort -k1,1 -k2,2n -k3,3n -k10,10n > /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_withMotifHits.bed
python /home/ikaplow/RegulatoryElementEvolutionProject/src/RegElEvoCode/selectSummitsNearMotifHits.py --peakMotifHitFileName /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_withMotifHits.bed --outputFileName /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs.bed
liftOver /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs.bed /home/ikaplow/RegulatoryElementEvolutionProject/LiftoverChains/mm10ToHg38.over.chain.gz /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_hg38.bed /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_unmapped.bed

# Obtaining /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MPRAOneDesign/rheMac8.livBr.humanCoordFirst.bed:
awk 'BEGIN{OFS="\t"} {print $7,$8,$9,$1,$2,$3,$4,$5,$6,$10,$11,$12,$13,$14,$15,$16,$17,$18,$19,$20,$21,$22,$23,$24,$25}' /home/apfennin/projects/striatumEnh/1exp/xComp/out16/resTab/rankedDiff.livBr.verm16_rheMac8.1.6.2mm.txt | sed '1d' > /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MPRAOneDesign/rheMac8.livBr.humanCoordFirst.bed

# Obtaining /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_hg38_activityMammals.bed:
awk 'BEGIN{OFS="\t"} $2 > 1' /home/apfennin/projects/striatumEnh/1exp/xComp/out16/resTab/rankedDiff.livBr.verm16_hg38.1.6.2mm.txt | awk 'BEGIN{OFS="\t"} $5 < 0.05' | cut -f7-9 | intersectBed -u -wa -a /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_hg38.bed -b stdin | intersectBed -wa -wb -b /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MPRAOneDesign/rheMac8.livBr.humanCoordFirst.bed -a stdin | awk 'BEGIN{OFS="\t"} $8 > 1' | awk 'BEGIN{OFS="\t"} $11 < 0.05' | cut -f1-3 | sort -u -k1,1 -k2,2n -k3,3n > /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_hg38_activityMammals.bed

# Obtaining /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_hg38_activityMouse.bed:
awk 'BEGIN{OFS="\t"} $2 <= 0' /home/apfennin/projects/striatumEnh/1exp/xComp/out16/resTab/rankedDiff.livBr.verm16_hg38.1.6.2mm.txt | cut -f7-9 | intersectBed -u -wa -a /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_hg38.bed -b stdin | intersectBed -wa -wb -b /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MPRAOneDesign/rheMac8.livBr.humanCoordFirst.bed -a stdin | awk 'BEGIN{OFS="\t"} $8 <= 0' | cut -f1-3 | sort -u -k1,1 -k2,2n -k3,3n > /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_hg38_activityMouse.bed

# Obtaining /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifsAndAllBrainDownGenes_hg38.bed:
python /home/ikaplow/RegulatoryElementEvolutionProject/src/getEnsemblCoordinatesFromGff3.py --gff3FileName /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/HumanGenome/gencode.v27.annotation.protTranscript.gff3 --ensemblGenesFileName /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MPRAOneDesign/Brawand2011Data/allBrainDown --outputFileName /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MPRAOneDesign/Brawand2011Data/allBrainDownGeneCoordinates.bed
sort -u -k1,1 -k2,2n -k3,3n /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MPRAOneDesign/Brawand2011Data/allBrainDownGeneCoordinates.bed > /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MPRAOneDesign/Brawand2011Data/allBrainDownGeneCoordinates_sorted.bed
awk 'BEGIN{OFS="\t"}{print $1,$2-1,$2}' /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MPRAOneDesign/Brawand2011Data/allBrainDownGeneCoordinates_sorted.bed > /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MPRAOneDesign/Brawand2011Data/allBrainDownGeneTSSs.bed
sort -u -k1,1 -k2,2n -k3,3n /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_hg38.bed | closestBed -d -t first -a stdin -b /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MPRAOneDesign/Brawand2011Data/allBrainDownGeneTSSs.bed | awk 'BEGIN{OFS="\t"} $7 < 100000' | cut -f1-3 | sort -u -k1,1 -k2,2n -k3,3n > /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifsAndAllBrainDownGenes_hg38.bed

# Obtaining /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_conserved_hg38.bed:
liftOver /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_hg38.bed /home/apfennin/tools/chain/hg38ToHg19.over.chain.gz /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_hg19.bed /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_unmappedHg19.bed
liftOver /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_hg19.bed /home/apfennin/tools/chain/hg19ToTaeGut2.over.chain.gz /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_taeGut2.bed /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_unmappedTaeGut2.bed
liftOver /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_taeGut2.bed /home/apfennin/tools/chain/taeGut2ToHg19.over.chain.gz /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_conserved_hg19.bed /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_conserved_unmappedHg19.bed
liftOver /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_conserved_hg19.bed /home/apfennin/tools/chain/hg19ToHg38.over.chain.gz /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_conserved_hg38.bed /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_conserved_unmappedHg38.bed

# Obtaining /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_plusMinus60bp_mm10_activityMammalsOrMouseOrNearBrainDownGenesOrConserved_noExon_Mef2cMotifHitsBrainEnhancerBackground:
cat /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_hg38_activityMammals.bed /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_hg38_activityMouse.bed /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifsAndAllBrainDownGenes_hg38.bed /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_conserved_hg38.bed | sort -u -k1,1 -k2,2n -k3,3n > /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_hg38_activityMammalsOrMouseOrNearBrainDownGenesOrConserved.bed
liftOver /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_hg38_activityMammalsOrMouseOrNearBrainDownGenesOrConserved.bed /home/apfennin/tools/chain/hg38ToMm10.over.chain.gz /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_mm10_activityMammalsOrMouseOrNearBrainDownGenesOrConserved.bed /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_unmapped_activityMammalsOrMouseOrNearBrainDownGenesOrConserved.bed
zcat /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MouseGenome/gencode.vM15.annotation.gff3.gz | grep exon | grep -v pseudogene | cut -f1,4,5 | subtractBed -A -a /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_mm10_activityMammalsOrMouseOrNearBrainDownGenesOrConserved.bed -b stdin > /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_mm10_activityMammalsOrMouseOrNearBrainDownGenesOrConserved_noExon.bed
awk 'BEGIN{OFS="\t"}{print $1,$3-60,$3+60}' /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_mm10_activityMammalsOrMouseOrNearBrainDownGenesOrConserved_noExon.bed | fastaFromBed -fi /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MouseGenome/mm10.fa -bed stdin -fo /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_plusMinus60bp_mm10_activityMammalsOrMouseOrNearBrainDownGenesOrConserved_noExon.fa
fimo --bgfile /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MPRAOneDesign/mouseBrainMinusGpeLiver_crossAge_allBrainMinusGpeUp_enhancer_backgroundDinuc.txt --max-stored-scores 5000000 --motif M4467_1.02 --thresh 0.001 --o /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_plusMinus60bp_mm10_activityMammalsOrMouseOrNearBrainDownGenesOrConserved_noExon_Mef2cMotifHitsBrainEnhancerBackground /home/ikaplow/RegulatoryElementEvolutionProject/MotifData/motif_databases/CIS-BP/HumanMacaqueMouseFinchFilt.meme /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_plusMinus60bp_mm10_activityMammalsOrMouseOrNearBrainDownGenesOrConserved_noExon.fa

# Obtaining /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_plusMinus60bp_mm10_activityMammalsOrMouseOrNearBrainDownGenesOrConserved_noExon_shuffledAllMef2cMotifs_bestScrambleSeqs.fa
python /home/ikaplow/RegulatoryElementEvolutionProject/src/RegElEvoCode/convertFIMOToMotifHitBed.py --fimoFileName /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_plusMinus60bp_mm10_activityMammalsOrMouseOrNearBrainDownGenesOrConserved_noExon_Mef2cMotifHitsBrainEnhancerBackground/fimo.txt --nameCol 1 --pValCutoff 0.001 --outputFileName /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_plusMinus60bp_mm10_activityMammalsOrMouseOrNearBrainDownGenesOrConserved_noExon_Mef2cMotifHitsBrainEnhancerBackground/fimoAll.bed
fastaFromBed -fi /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MouseGenome/mm10.fa -bed /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_plusMinus60bp_mm10_activityMammalsOrMouseOrNearBrainDownGenesOrConserved_noExon_Mef2cMotifHitsBrainEnhancerBackground/fimoAll.bed -fo /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_plusMinus60bp_mm10_activityMammalsOrMouseOrNearBrainDownGenesOrConserved_noExon_Mef2cMotifHitsBrainEnhancerBackground/fimoAll.fa
fasta-shuffle-letters -copies 500 -dna /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_plusMinus60bp_mm10_activityMammalsOrMouseOrNearBrainDownGenesOrConserved_noExon_Mef2cMotifHitsBrainEnhancerBackground/fimoAll.fa /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_plusMinus60bp_mm10_activityMammalsOrMouseOrNearBrainDownGenesOrConserved_noExon_Mef2cMotifHitsBrainEnhancerBackground/fimoAll_shuffled.fa
python /home/ikaplow/RegulatoryElementEvolutionProject/src/RegElEvoCode/replaceFIMOMotifHitWithShuffledSequence.py --fastaFileName /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_plusMinus60bp_mm10_activityMammalsOrMouseOrNearBrainDownGenesOrConserved_noExon.fa --fimoFileName /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_plusMinus60bp_mm10_activityMammalsOrMouseOrNearBrainDownGenesOrConserved_noExon_Mef2cMotifHitsBrainEnhancerBackground/fimo.txt --shuffledFastaFileName /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_plusMinus60bp_mm10_activityMammalsOrMouseOrNearBrainDownGenesOrConserved_noExon_Mef2cMotifHitsBrainEnhancerBackground/fimoAll_shuffled.fa --pValCutoff 0.001 --outputFileName /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_plusMinus60bp_mm10_activityMammalsOrMouseOrNearBrainDownGenesOrConserved_noExon_shuffledAllMef2cMotifs.fa
fimo --bgfile /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MPRAOneDesign/mouseBrainMinusGpeLiver_crossAge_allBrainMinusGpeUp_enhancer_backgroundDinuc.txt --max-stored-scores 5000000 --thresh 0.001 --o /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_plusMinus60bp_mm10_activityMammalsOrMouseOrNearBrainDownGenesOrConserved_noExon_shuffledAllMef2cMotifs_MotifHits /home/ikaplow/RegulatoryElementEvolutionProject/MotifData/motif_databases/CIS-BP/HumanMacaqueMouseFinchFilt.meme /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_plusMinus60bp_mm10_activityMammalsOrMouseOrNearBrainDownGenesOrConserved_noExon_shuffledAllMef2cMotifs.fa
python /home/ikaplow/RegulatoryElementEvolutionProject/src/RegElEvoCode/convertFIMOToMotifHitBed.py --fimoFileName /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_plusMinus60bp_mm10_activityMammalsOrMouseOrNearBrainDownGenesOrConserved_noExon_shuffledAllMef2cMotifs_MotifHits/fimo.txt --nameCol 1 --pValCutoff 0.001 --nameCol 2 --scoreCol 7 --outputFileName /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_plusMinus60bp_mm10_activityMammalsOrMouseOrNearBrainDownGenesOrConserved_noExon_shuffledAllMef2cMotifs_MotifHits/fimo.bed
python /home/ikaplow/RegulatoryElementEvolutionProject/src/RegElEvoCode/convertFIMOToMotifHitBed.py --fimoFileName /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_plusMinus60bp_mm10_activityMammalsOrMouseOrNearBrainDownGenesOrConserved_noExon_Mef2cMotifHitsBrainEnhancerBackground/fimo.txt --nameCol 1 --pValCutoff 0.0001 --outputFileName /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_plusMinus60bp_mm10_activityMammalsOrMouseOrNearBrainDownGenesOrConserved_noExon_Mef2cMotifHitsBrainEnhancerBackground/fimoAllStrict.bed
intersectBed -wa -u -a /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_plusMinus60bp_mm10_activityMammalsOrMouseOrNearBrainDownGenesOrConserved_noExon_shuffledAllMef2cMotifs_MotifHits/fimo.bed -b /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_plusMinus60bp_mm10_activityMammalsOrMouseOrNearBrainDownGenesOrConserved_noExon_Mef2cMotifHitsBrainEnhancerBackground/fimoAllStrict.bed | sort -u -k4,4 -k5,5g -k1,1 -k2,2n -k3,3n > /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_plusMinus60bp_mm10_activityMammalsOrMouseOrNearBrainDownGenesOrConserved_noExon_shuffledAllMef2cMotifs_MotifHits/fimo_originalMef2cMotifHit.bed
python /home/ikaplow/RegulatoryElementEvolutionProject/src/RegElEvoCode/chooseBestScramble.py --fimoBedFileName /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_plusMinus60bp_mm10_activityMammalsOrMouseOrNearBrainDownGenesOrConserved_noExon_shuffledAllMef2cMotifs_MotifHits/fimo_originalMef2cMotifHit.bed --outputFileName /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_plusMinus60bp_mm10_activityMammalsOrMouseOrNearBrainDownGenesOrConserved_noExon_shuffledAllMef2cMotifs_bestScrambles.txt
python /home/ikaplow/RegulatoryElementEvolutionProject/src/RegElEvoCode/filterFastaFile.py --fastaFileName /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_plusMinus60bp_mm10_activityMammalsOrMouseOrNearBrainDownGenesOrConserved_noExon_shuffledAllMef2cMotifs.fa --fastaIDsFileName /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_plusMinus60bp_mm10_activityMammalsOrMouseOrNearBrainDownGenesOrConserved_noExon_shuffledAllMef2cMotifs_bestScrambles.txt --outputFileName /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_plusMinus60bp_mm10_activityMammalsOrMouseOrNearBrainDownGenesOrConserved_noExon_shuffledAllMef2cMotifs_bestScrambleSeqs.fa

# Obtaining /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_plusMinus60bp_mm10_activityMammalsOrMouseOrNearBrainDownGenesOrConserved_noExon_shuffledAllMef2cMotifsPlusMinus5bp_bestScrambleSeqs.fa
awk 'BEGIN{OFS="\t"} {print $1, $2-5, $3+5}' /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_plusMinus60bp_mm10_activityMammalsOrMouseOrNearBrainDownGenesOrConserved_noExon_Mef2cMotifHitsBrainEnhancerBackground/fimoAll.bed | fastaFromBed -fi /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MouseGenome/mm10.fa -bed stdin -fo /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_plusMinus60bp_mm10_activityMammalsOrMouseOrNearBrainDownGenesOrConserved_noExon_Mef2cMotifHitsBrainEnhancerBackground/fimoAllPlusMinus5bp.fa
fasta-shuffle-letters -copies 500 -dna /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_plusMinus60bp_mm10_activityMammalsOrMouseOrNearBrainDownGenesOrConserved_noExon_Mef2cMotifHitsBrainEnhancerBackground/fimoAllPlusMinus5bp.fa /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_plusMinus60bp_mm10_activityMammalsOrMouseOrNearBrainDownGenesOrConserved_noExon_Mef2cMotifHitsBrainEnhancerBackground/fimoAllPlusMinus5bp_shuffled.fa
fimo --bgfile /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MPRAOneDesign/mouseBrainMinusGpeLiver_crossAge_allBrainMinusGpeUp_enhancer_backgroundDinuc.txt --max-stored-scores 5000000 --thresh 0.001 --o /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_plusMinus60bp_mm10_activityMammalsOrMouseOrNearBrainDownGenesOrConserved_noExon_shuffledAllMef2cMotifsPlusMinus5bp_MotifHits /home/ikaplow/RegulatoryElementEvolutionProject/MotifData/motif_databases/CIS-BP/HumanMacaqueMouseFinchFilt.meme /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_plusMinus60bp_mm10_activityMammalsOrMouseOrNearBrainDownGenesOrConserved_noExon_shuffledAllMef2cMotifsPlusMinus5bp.fa
python /home/ikaplow/RegulatoryElementEvolutionProject/src/RegElEvoCode/convertFIMOToMotifHitBed.py --fimoFileName /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_plusMinus60bp_mm10_activityMammalsOrMouseOrNearBrainDownGenesOrConserved_noExon_shuffledAllMef2cMotifsPlusMinus5bp_MotifHits/fimo.txt --nameCol 1 --pValCutoff 0.001 --nameCol 2 --scoreCol 7 --outputFileName /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_plusMinus60bp_mm10_activityMammalsOrMouseOrNearBrainDownGenesOrConserved_noExon_shuffledAllMef2cMotifsPlusMinus5bp_MotifHits/fimo.bed
intersectBed -wa -u -a /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_plusMinus60bp_mm10_activityMammalsOrMouseOrNearBrainDownGenesOrConserved_noExon_shuffledAllMef2cMotifsPlusMinus5bp_MotifHits/fimo.bed -b /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_plusMinus60bp_mm10_activityMammalsOrMouseOrNearBrainDownGenesOrConserved_noExon_Mef2cMotifHitsBrainEnhancerBackground/fimoAllStrict.bed | sort -u -k4,4 -k5,5g -k1,1 -k2,2n -k3,3n > /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_plusMinus60bp_mm10_activityMammalsOrMouseOrNearBrainDownGenesOrConserved_noExon_shuffledAllMef2cMotifsPlusMinus5bp_MotifHits/fimo_originalMef2cMotifHit.bed
python /home/ikaplow/RegulatoryElementEvolutionProject/src/RegElEvoCode/chooseBestScramble.py --fimoBedFileName /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_plusMinus60bp_mm10_activityMammalsOrMouseOrNearBrainDownGenesOrConserved_noExon_shuffledAllMef2cMotifsPlusMinus5bp_MotifHits/fimo_originalMef2cMotifHit.bed --outputFileName /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_plusMinus60bp_mm10_activityMammalsOrMouseOrNearBrainDownGenesOrConserved_noExon_shuffledAllMef2cMotifsPlusMinus5bp_bestScrambles.txt
python /home/ikaplow/RegulatoryElementEvolutionProject/src/RegElEvoCode/filterFastaFile.py --fastaFileName /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_plusMinus60bp_mm10_activityMammalsOrMouseOrNearBrainDownGenesOrConserved_noExon_shuffledAllMef2cMotifsPlusMinus5bp.fa --fastaIDsFileName /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_plusMinus60bp_mm10_activityMammalsOrMouseOrNearBrainDownGenesOrConserved_noExon_shuffledAllMef2cMotifsPlusMinus5bp_bestScrambles.txt --outputFileName /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_plusMinus60bp_mm10_activityMammalsOrMouseOrNearBrainDownGenesOrConserved_noExon_shuffledAllMef2cMotifsPlusMinus5bp_bestScrambleSeqs.fa

# Obtaining /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_mm10_activityMammalsOrMouseOrNearBrainDownGenesOrConserved_closestGenes.bed
sort -u -k1,1 -k2,2n -k3,3n /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MouseGenome/gencode.vM15.annotation.protTranscript.geneNames.bed | closestBed -d -a /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_mm10_activityMammalsOrMouseOrNearBrainDownGenesOrConserved_sorted.bed -b stdin > /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_mm10_activityMammalsOrMouseOrNearBrainDownGenesOrConserved_closestGenes.bed

# Obtaining /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_mm10_activityMammalsOrMouseOrNearBrainDownGenesOrConserved_noExon_DSBDown.bed:
liftOver /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/DNADamageData/tab_decreasing.txt /home/ikaplow/RegulatoryElementEvolutionProject/LiftoverChains/mm9ToMm10.over.chain.gz /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/DNADamageData/tab_decreasing_mm10.bed /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/DNADamageData/tab_decreasing_unmapped.bed
intersectBed -u -wa -a /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_mm10_activityMammalsOrMouseOrNearBrainDownGenesOrConserved_noExon.bed -b /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/DNADamageData/tab_decreasing_mm10.bed | sort -k1,1 -k2,2n -k3,3n > /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_mm10_activityMammalsOrMouseOrNearBrainDownGenesOrConserved_noExon_DSBDown.bed

# Obtaining /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_mm10_activityMammalsOrMouseOrNearBrainDownGenesOrConserved_noExon_inLongPeak.bed:
intersectBed -wa -wb -a /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_mm10_activityMammalsOrMouseOrNearBrainDownGenesOrConserved_noExon.bed -b /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MouseDNase/Cortex_AgeB_ATAC_out/peak/macs2/idr/optimal_set/Cortex_AgeB_ATAC_out_ppr.IDR0.1.filt.narrowPeak.gz | awk '$6-$5 > 1000' | sort -u -k1,1 -k2,2n -k3,3n > /projects/pfenninggroup/machineLearningForComputationalBiology/regElEvoGrant/MEF2CData/peak/spp/idr/optimal_set/MEF2CData_rep1-pr.IDR0.05.filt_enhancer_summitsNearMotifs_mm10_activityMammalsOrMouseOrNearBrainDownGenesOrConserved_noExon_inLongPeak.bed